pkg/workloadmeta/store.go

```diff
@@ -0,0 +1,382 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+package workloadmeta
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/DataDog/datadog-agent/pkg/errors"
+	"github.com/DataDog/datadog-agent/pkg/status/health"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+	"github.com/DataDog/datadog-agent/pkg/util/retry"
+)
+
+var (
+	globalStore *Store
+	initOnce    sync.Once
+)
+
+const (
+	retryCollectorInterval = 30 * time.Second
+	pullCollectorInterval  = 5 * time.Second
+	eventBundleChTimeout   = 1 * time.Second
+	eventChBufferSize      = 50
+)
+
+type subscriber struct {
+	name   string
+	ch     chan EventBundle
+	filter *Filter
+}
+
+// Store is a central storage of metadata about workloads. A workload is any
+// unit of work being done by a piece of software, like a process, a container,
+// a kubernetes pod, or a task in any cloud provider.
+type Store struct {
+	storeMut sync.RWMutex
+	store    map[Kind]map[string]Entity
+
+	subscribersMut sync.RWMutex
+	subscribers    []subscriber
+
+	candidates map[string]Collector
+	collectors map[string]Collector
+
+	eventCh chan []Event
+}
+
+// NewStore creates a new workload metadata store, building a new instance of
+// each registered collector. Call Run to start the store and its collectors.
+func NewStore() *Store {
+	candidates := make(map[string]Collector)
+	for id, c := range collectorCatalog {
+		candidates[id] = c()
+	}
+
+	return &Store{
+		store:       make(map[Kind]map[string]Entity),
+		subscribers: []subscriber{},
+
+		candidates: candidates,
+		collectors: make(map[string]Collector),
+		eventCh:    make(chan []Event, eventChBufferSize),
+	}
+}
+
+// Run starts the workload metadata store.
+func (s *Store) Run(ctx context.Context) {
```

Comment:
Only a semantic comment.

I'm wondering if we should rename this function `Start` because we create a goroutine inside this function to run on the background.

so for me:
```golang

store.Start(ctx) # it starts the store on the background. 

store.Run(ctx) # keep running, only exit when it is done or ctx expired.
# can be also use like:
go store.Run(ctx) # with an explicit `goroutine creation`
```